packages/Meshfree/src/DTK_MovingLeastSquaresOperator_def.hpp

```diff
@@ -0,0 +1,139 @@
+
+#ifndef DTK_MOVING_LEAST_SQUARES_OPERATOR_DEF_HPP
+#define DTK_MOVING_LEAST_SQUARES_OPERATOR_DEF_HPP
+
+#include <DTK_DBC.hpp>
+#include <DTK_DetailsMovingLeastSquaresOperatorImpl.hpp>
+#include <DTK_DetailsNearestNeighborOperatorImpl.hpp> // makeDistributedSearchTree, fetch
+#include <DTK_DistributedSearchTree.hpp>
+
+namespace DataTransferKit
+{
+
+template <typename DeviceType, typename CompactlySupportedRadialBasisFunction,
+          typename PolynomialBasis>
+MovingLeastSquaresOperator<DeviceType, CompactlySupportedRadialBasisFunction,
+                           PolynomialBasis>::
+    MovingLeastSquaresOperator(
+        Teuchos::RCP<Teuchos::Comm<int> const> const &comm,
+        Kokkos::View<Coordinate const **, DeviceType> source_points,
+        Kokkos::View<Coordinate const **, DeviceType> target_points )
+    : _comm( comm )
+    , _n_source_points( source_points.extent( 0 ) )
+    , _offset( "offset" )
+    , _ranks( "ranks" )
+    , _indices( "indices" )
+    , _coeffs( "polynomial_coefficients" )
+{
+    DTK_REQUIRE( source_points.extent_int( 1 ) ==
+                 target_points.extent_int( 1 ) );
+    // FIXME for now let's assume 3D
+    DTK_REQUIRE( source_points.extent_int( 1 ) == 3 );
+
+    // Build distributed search tree over the source points.
+    auto search_tree = Details::NearestNeighborOperatorImpl<
+        DeviceType>::makeDistributedSearchTree( _comm, source_points );
+    DTK_CHECK( !search_tree.empty() );
+
+    // For each target point, query the n_neighbors points closest to the
+    // target.
+    auto queries =
+        Details::MovingLeastSquaresOperatorImpl<DeviceType>::makeKNNQueries(
+            target_points, PolynomialBasis::size() );
+
+    // Perform the actual search.
+    search_tree.query( queries, _indices, _offset, _ranks );
+
+    // Retrieve the coordinates of all source points that met the predicates.
+    // NOTE: This is the last collective.
+    source_points = Details::NearestNeighborOperatorImpl<DeviceType>::fetch(
+        _comm, _ranks, _indices, source_points );
+
+    // Transform source points
+    source_points = Details::MovingLeastSquaresOperatorImpl<
+        DeviceType>::transformSourceCoordinates( source_points, _offset,
+                                                 target_points );
+    target_points = Kokkos::View<double **, DeviceType>( "empty" );
+
+    // Build P (vandermonde matrix)
+    // P is a single 1D storage for multiple P_i matrices. Each matrix is of
+    // size (#source_points_for_specific_target_point, basis_size)
+    auto p =
+        Details::MovingLeastSquaresOperatorImpl<DeviceType>::computeVandermonde(
+            source_points, PolynomialBasis() );
+
+    // To build the radial basis function, we need to define the radius of the
+    // radial basis function. Since we use kNN, we need to compute the radius.
+    auto radius =
+        Details::MovingLeastSquaresOperatorImpl<DeviceType>::computeRadius(
+            source_points, _offset );
```

Comment:
Comment that you are not passing the target points to compute the radii because of the coordinate transformation.